app/services/structured_fact_importer.rb

```diff
@@ -0,0 +1,37 @@
+class StructuredFactImporter < FactImporter
+  def normalize(facts)
+    # Remove empty values first, so nil facts added by normalize_recurse imply compose
+    facts = facts.select { |k, v| v.present? }
+    normalize_recurse({}, facts)
+  end
+
+  # expand {'a' => {'b' => 'c'}} to {'a' => nil, 'a::b' => 'c'}
```

Comment:
Kind of the same as https://github.com/theforeman/foreman_ansible/blob/master/app/services/foreman_ansible/fact_sparser.rb  https://github.com/theforeman/foreman_chef/blob/1087d8a517b741f7b067d4d5d742593beaf5039b/app/models/foreman_chef/fact_importer.rb#L110 or https://github.com/theforeman/foreman_salt/blob/1ceeece560af143bb90b211ef8404c18537e3519/app/services/foreman_salt/fact_importer.rb#L92

As far as I can see the only difference is that this code generates a `nil` key to indicate facts composition. I'll look a bit further to see if that works for the other plugins as ideally these plugins can reutilize this 'hash sparser' from core